### PR TITLE
SV-COMP 2018: Rebase of Tautschnig's branch/PR `sv-comp-2017` on `cbmc/develop`

### PR DESCRIPTION
**Pull request by marek-trtik**
@tautschnig , @peterschrammel @martin-cs : Please check the correctness of the rebase. Thank you.

!!! IMPORTANT: I ask other reviewers to postpone their reviews till the PR is fixed (NOTE: I had to create the PR in order to fix issues in Travis and mainly AppVeyor). Once the PR is ready for reviews I'll post an explicit message to this PR. Thank you for understanding. !!!

This is the rebased PR: https://github.com/diffblue/cbmc/pull/363

### CODE REVIEW

**Comment by peterschrammel** [comment]
@marek-trtik, unfortunately, the minisat download is hard-coded in .travis.yml: https://travis-ci.org/diffblue/cbmc/jobs/293594798#L729
You'll have to hack https://github.com/diffblue/cbmc/blob/develop/.travis.yml#L239 to make this build on CI.

**Comment by marek-trtik** [comment]
Fixed. Thank you Petr!

File: src/solvers/flattening/bv_pointers.cpp [hunk 1]
```diff
@@ -784,6 +773,78 @@ void bv_pointerst::do_postponed(
...
+    bvt disj;
+    disj.reserve(objects.size());
+
+    bvt saved_bv=postponed.op;
+    saved_bv.erase(saved_bv.begin(), saved_bv.begin()+offset_bits);
+
+    bvt invalid_bv, null_bv;
+    encode(pointer_logic.get_invalid_object(), invalid_bv);
+    invalid_bv.erase(invalid_bv.begin(), invalid_bv.begin()+offset_bits);
+    encode(pointer_logic.get_null_object(),    null_bv);
```
Comment (romainbrenguier) [inline src/solvers/flattening/bv_pointers.cpp]: weird spacing

File: src/solvers/flattening/bv_pointers.cpp [hunk 2]
```diff
@@ -439,6 +425,9 @@ bvt bv_pointerst::convert_pointer_type(const exprt &expr)
   {
     return SUB::convert_byte_extract(to_byte_extract_expr(expr));
   }
+  else if(expr.id()==ID_byte_update_little_endian ||
+          expr.id()==ID_byte_update_big_endian)
+    throw "byte-wise updates of pointers are unsupported";
```
Comment (romainbrenguier) [inline src/solvers/flattening/bv_pointers.cpp]: not sure we should use `throw` if there is no corresponding `catch`

File: src/util/guard.cpp [hunk 3]
```diff
@@ -95,9 +95,9 @@ guardt &operator -= (guardt &g1, const guardt &g2)
   if(g1.id()!=ID_and || g2.id()!=ID_and)
     return g1;
 
-  sort_and_join(g1);
+  // sort_and_join(g1);
```
Comment (romainbrenguier) [inline src/util/guard.cpp]: remove or use `#if 0` and explain

File: src/util/guard.cpp [hunk 4]
```diff
@@ -108,10 +108,12 @@ guardt &operator -= (guardt &g1, const guardt &g2)
       it2!=op2.end();
       ++it2)
   {
-    while(it1!=op1.end() && *it1<*it2)
-      ++it1;
+    /*while(it1!=op1.end() && *it1<*it2)
```
Comment (romainbrenguier) [inline src/util/guard.cpp]: remove

File: src/util/guard.cpp [hunk 5]
```diff
@@ -145,9 +147,9 @@ guardt &operator |= (guardt &g1, const guardt &g2)
   }
 
   // find common prefix
-  sort_and_join(g1);
+  // sort_and_join(g1);
```
Comment (romainbrenguier) [inline src/util/guard.cpp]: remove or use `#if 0`

File: src/util/simplify_expr_boolean.cpp [hunk 6]
```diff
@@ -72,21 +72,60 @@ bool simplify_exprt::simplify_boolean(exprt &expr)
...
+
+      if(it->is_true())
+      {
+        erase=true;
+        negate=!negate;
+      }
+      else
+        erase=it->is_false();
+
+      if(erase)
```
Comment (romainbrenguier) [inline src/util/simplify_expr_boolean.cpp]: what about getting rid of the `erase` variable and replace this condition by `if(it->is_constant())`

File: src/util/simplify_utils.cpp [hunk 7]
```diff
@@ -119,7 +119,7 @@ static const struct saj_tablet &sort_and_join(
   return saj_table[i];
 }
 
-bool sort_and_join(exprt &expr)
+bool sort_and_join(exprt &expr, bool do_sort)
```
Comment (romainbrenguier) [inline src/util/simplify_utils.cpp]: Would it make more sense to have a `sort` and a `join` function to separate the two functionalities?

File: regression/cbmc-concurrency/thread_chain_posix2/test.desc [hunk 8]
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
Comment (romainbrenguier) [inline regression/cbmc-concurrency/thread_chain_posix2/test.desc]: Why is this broken? Is there an issue raised to fix it?

File: src/analyses/goto_check.cpp [hunk 9]
```diff
@@ -1655,10 +1656,37 @@ void goto_checkt::goto_check(
...
+      {
+        INVARIANT(i.code.operands().size()==1, "The num-operands must be 1.");
+        const symbol_exprt &variable=to_symbol_expr(i.code.op0());
+
+        // is it dirty?
+        if(local_bitvector_analysis->dirty(variable))
+        {
+          // reset the dead marker
+          goto_programt::targett t=new_code.add_instruction(ASSIGN);
+          exprt address_of_expr=address_of_exprt(variable);
```
Comment (romainbrenguier) [inline src/analyses/goto_check.cpp]: `address_of_exprt address_of_expr(variable);` ?

File: src/analyses/goto_check.cpp [hunk 10]
```diff
@@ -1655,10 +1656,37 @@ void goto_checkt::goto_check(
...
+          exprt lhs=ns.lookup(CPROVER_PREFIX "dead_object").symbol_expr();
+          if(!base_type_eq(lhs.type(), address_of_expr.type(), ns))
+            address_of_expr.make_typecast(lhs.type());
+          exprt rhs=
+            if_exprt(
```
Comment (romainbrenguier) [inline src/analyses/goto_check.cpp]: `if_exprt rhs`

File: src/ansi-c/expr2c.cpp [hunk 11]
```diff
@@ -699,6 +699,7 @@ std::string expr2ct::convert_struct_type(
   if(tag!="")
     dest+=" "+id2string(tag);
 
+#if 0
```
Comment (romainbrenguier) [inline src/ansi-c/expr2c.cpp]: why is this commented out?

File: src/ansi-c/expr2c.cpp [hunk 12]
```diff
@@ -1115,17 +1117,20 @@ std::string expr2ct::convert_pointer_object_has_type(
...
-  if(src.operands().size()!=1)
+  if(src.operands().size()!=2)
     return convert_norep(src, precedence);
 
   unsigned p0;
   std::string op0=convert_with_precedence(src.op0(), p0);
 
-  std::string dest="MALLOC";
+  unsigned p1;
+  std::string op1=convert_with_precedence(src.op0(), p1);
```
Comment (romainbrenguier) [inline src/ansi-c/expr2c.cpp]: const ?

File: src/analyses/goto_check.cpp [hunk 13]
```diff
@@ -1655,10 +1656,37 @@ void goto_checkt::goto_check(
...
+          t->code=code_assignt(lhs, rhs);
+          t->code.add_source_location()=i.source_location;
+        }
+      }
+    }
     else if(i.is_end_function())
     {
-      if(i.function==goto_functionst::entry_point() &&
-         enable_memory_leak_check)
+      if(i.function==goto_functionst::entry_point() && enable_memory_leak_check)
```
Comment (romainbrenguier) [inline src/analyses/goto_check.cpp]: this change is not in the right commit

File: .travis.yml [hunk 14]
```diff
@@ -236,10 +236,10 @@ jobs:
 
 install:
   - ccache --max-size=1G
-  - make -C src minisat2-download
+  - make -C src glucose-download
   - make -C src/ansi-c library_check
-  - make -C src "CXX=${COMPILER}" "CXXFLAGS=-Wall -Werror -pedantic -O2 -g ${EXTRA_CXXFLAGS}" -j2
-  - make -C src "CXX=${COMPILER}" "CXXFLAGS=-Wall -Werror -pedantic -O2 -g ${EXTRA_CXXFLAGS}" -j2 clobber.dir memory-models.dir
```
Comment (romainbrenguier) [inline .travis.yml]: I don't think that's a good idea to remove these flags

**Review by romainbrenguier — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 15]
```diff
@@ -211,7 +211,7 @@ jobs:
             name: "diffblue/cbmc"
             description: "Travis build of ${TRAVIS_COMMIT}"
           notification_email: "coverity-scan@diffblue.com"
-          build_command_prepend: "make -C src minisat2-download"
+          build_command_prepend: "make -C src glucose-download"
```
Comment (thk123) [inline .travis.yml]: This seems like a surprising change - what are the implications/justifications for this swap?

File: src/config.inc [hunk 16]
```diff
@@ -5,7 +5,7 @@ BUILD_ENV = AUTO
 ifeq ($(BUILD_ENV),MSVC)
   #CXXFLAGS += /Wall /WX
 else
-  CXXFLAGS += -Wall -pedantic -Werror
+  CXXFLAGS += -Wall -Wno-parentheses
```
Comment (thk123) [inline src/config.inc]: This shouldn't change

**Review by thk123 — COMMENTED** [review]
I don't think this is ready for review - though I agree with 
@romainbrenguier's comments. Could you comment when it needs another review.

**Comment by tautschnig** [comment]
I'm wondering how to proceed with #982 and #731, which contain a collection of the patches in this PR. Pushing on them would mean we get duplicate work through this one.

**Comment by marek-trtik** [comment]
@thk123 : Yes, the PR is NOT ready for a review yet. I'll post here an explicit message, when the PR is ready.

**Comment by marek-trtik** [comment]
@tautschnig : According to your response to @lucasccordeiro 's email I believe that #982 and #731 are (some of) those PRs factored out from #363. I think that it is still true that #363 contains all patches for SV_COMP we currently have. Also, everybody I asked told me that I should work with #363.

The goal with #363 was to prepare two branches, both rebased on the latest `develop`. The first branch should contain changes which should go to `develop` right now, and the other consisting of only SV-COMP related features. The second branch is supposed to be rebased on top of the first one. The second branch will also go to `develop`, but not now (after SV-COMP deadline). We will introduce a macro `SV_COMP` and all that code will be compiled only when the macro is defined. 

Both branches are already created and they are rebased on latest `develop`. The first branch is already PRed (this PR). There is only one remaining thing to be resolved - fixing regression tests on Windows (on Linux build all pass). 

I did not consider #982 nor #731 in this process. I am not sure whether it make a sense to use them at this point.

**Comment by marek-trtik** [comment]
@tautschnig : I could merge current states of #982 and #731 into this PR.

What solution would you propose?

**Comment by peterschrammel** [comment]
@tautschnig, Marek is fixing a long-on-Windows-is-32bits bug in these assumptions: https://github.com/tautschnig/cbmc/blob/799446487a95302106bbe8d1a9de9a88d46eceee/src/ansi-c/library/new.c#L14
Is it worth fixing them or have they become obsolete in the meanwhile?

**Comment by tautschnig** [comment]
@marek-trtik 

> @tautschnig : I could merge current states of #982 and #731 into this PR.
What solution would you propose?

If you could replace all commits with the same subject with their versions from #982 and #731 this would indeed be nice!

@peterschrammel 
> @tautschnig, Marek is fixing a long-on-Windows-is-32bits bug in these assumptions: 

I've digged into my various pull requests, but it seems I've dropped this everywhere. One problem certainly is that the object:offset encoding is now configurable, hence any such check is somewhat questionable. Hopefully #1086 is something I'll find time for very soon...

**Comment by marek-trtik** [comment]
@tautschnig : Yes, I'll try to take versions of the corresponding commits from #982 and #731 (however, I'll keep fixes (CPPLINT/clang-formater) I made). So, it means, that I'll perform merge of the related commits instead of replacement.

Regarding the long-on-Windows-is-32bits issue: If I understood correctly #1086 should solve (avoid) the issue (those assumptions won't be needed anymore). However, this PR cannot be merged without the issue is solved (one way of the other). The question is whether we want to wait with merge of this PR till #1086 is merged.

**Comment by tautschnig** [comment]
> The question is whether we want to wait with merge of this PR till #1086 is merged.

Let's try to avoid dependencies wherever possible, so: don't wait. A hack-fix should do for the moment.

**Comment by tautschnig** [comment]
Note that I have rebased #982 to make picking a bit easier.

**Comment by marek-trtik** [comment]
I fixed the issue long-on-Windows-is-32bits.

**Comment by marek-trtik** [comment]
@tautschnig : Thanks for the rebase. It will certainly help. However, I see there are still CPPLINT/clang-formater issues. I'll fixed them during the merge of corresponding commits.

**Comment by marek-trtik** [comment]
This PR now comprises #982 and #731. Also the assumptions (those with the  long-on-Windows-is-32bits issue) were moved into the SV-COMP specific branch:
https://github.com/marek-trtik/cbmc/tree/sv-comp-2017-specific
which is rebased on top of the branch of this PR. However, 2 regression tests fails on the `sv-comp-2017-specific` branch:
```
regression/cbmc-java/generics_symtab1/test.desc
regression/cbmc-java/json_trace2/test.desc
```
I'll investigate that (but this is irrelevant to this PR).

File: .travis.yml (hunk 14 above)
Comment (marek-trtik) [reply .travis.yml]: Sure, it is gone.

File: regression/cbmc-concurrency/thread_chain_posix2/test.desc (hunk 8 above)
Comment (marek-trtik) [reply regression/cbmc-concurrency/thread_chain_posix2/test.desc]: The change discarded.

File: src/analyses/goto_check.cpp (hunk 9 above)
Comment (marek-trtik) [reply src/analyses/goto_check.cpp]: Fixed.

File: src/analyses/goto_check.cpp (hunk 10 above)
Comment (marek-trtik) [reply src/analyses/goto_check.cpp]: Fixed.

File: src/analyses/goto_check.cpp (hunk 13 above)
Comment (marek-trtik) [reply src/analyses/goto_check.cpp]: Reverted back.

File: src/ansi-c/expr2c.cpp (hunk 11 above)
Comment (marek-trtik) [reply src/ansi-c/expr2c.cpp]: Reverted back.

File: src/ansi-c/expr2c.cpp (hunk 12 above)
Comment (marek-trtik) [reply src/ansi-c/expr2c.cpp]: Fixed.

File: src/solvers/flattening/bv_pointers.cpp (hunk 2 above)
Comment (marek-trtik) [reply src/solvers/flattening/bv_pointers.cpp]: Sure. However, there are already 656 such `throw "...` statements in our code base, e.g.:
https://github.com/diffblue/cbmc/blob/develop/src/analyses/ai.cpp#L466

File: src/solvers/flattening/bv_pointers.cpp (hunk 1 above)
Comment (marek-trtik) [reply src/solvers/flattening/bv_pointers.cpp]: Fixed.

File: src/util/guard.cpp (hunk 3 above)
Comment (marek-trtik) [reply src/util/guard.cpp]: Fixed.

File: src/util/guard.cpp (hunk 4 above)
Comment (marek-trtik) [reply src/util/guard.cpp]: Fixed.

File: src/util/guard.cpp (hunk 5 above)
Comment (marek-trtik) [reply src/util/guard.cpp]: Fixed.

File: src/util/simplify_expr_boolean.cpp (hunk 6 above)
Comment (marek-trtik) [reply src/util/simplify_expr_boolean.cpp]: Fixed.

File: src/util/simplify_utils.cpp (hunk 7 above)
Comment (marek-trtik) [reply src/util/simplify_utils.cpp]: Reverted back.

**Comment by marek-trtik** [comment]
!!! To all reviewers: This PR is ready for your reviews !!!

File: .travis.yml (hunk 15 above)
Comment (marek-trtik) [reply .travis.yml]: Reverted back.

File: src/config.inc (hunk 16 above)
Comment (marek-trtik) [reply src/config.inc]: Reverted back.

**Comment by tautschnig** [comment]
Breaking up into series of 19 pull requests starting from #1990 all the way up to #2008.

**Pull request closed without merging**